### ISSUE TITLE
Discuss: Handling updates using the reconciler framwork

### ISSUE THREAD

**Issue by rossf7** [giantswarm/operatorkit#58]
This issue is to get consensus on how we handle updates using the reconciler framework.

### Idempotency

* Operators should be idempotent. So should the addFunc and updateFunc implementations be the same?
* Another benefit is simplicity. Only the addFunc and deleteFunc logic needs to implemented. ~~The update logic is more complex and needs to be implemented in the framework and the operators.~~

### Ordering of updates

* Due to the resync period and outside factors like the operator being restarted we cannot guarantee the ordering of add and update events.
* Since the ordering is not guaranteed can we rely on using the updateFunc?

**Comment by rossf7** [comment]
@kopiczko @xh3b4sd @fgimenez Here is the discussion issue please add your opinions.

**Comment by xh3b4sd** [comment]
Towards your comments/questions. 

> Operators should be idempotent. So should the addFunc and updateFunc implementations be the same?

Yes, calling these functions multiple times should always result in (the same) predicable state. This implies that each step implementation of a resource that has write permissions (ProcessCreate, ProcessDelete, ProcessUpdate) has to be idempotent. 

> Only the addFunc and deleteFunc logic needs to implemented.

IMO this depends on the operator and the resources it implements. There are indeed operators who do not care about updates because the resources the operator cares about have only binary states and cannot be updated. In this case a resource is either there or not there. In case it is there it has a specific state that does not change. E.g. there could be an backup operator that only creates backups. Backups are not updated, they are just created. 

> The update logic is more complex and needs to be implemented in the framework and the operators.

I would disagree with this based on this single sentence. Please provide more information why you think so and at best some examples. 

> Due to the resync period and outside factors like the operator being restarted we cannot guarantee the ordering of add and update events.

That sounds like an insufficiency of the k8s client. You cannot update something that was not added before. It should actually be guaranteed that the add function is run before the update function. We should get in touch with the community if this is really an issue.

---

Towards my thinking of how updates should be implemented. We have the following cases. 

- Operators do not care about updates. They only implement add and delete. See my backup operator example above. 
- Operators care about "updates" without implementing the update function. This can be possible when an operator cares about multiple entities of the same resource which has a binary state. Then entities are created and deleted. Due to the changing resources the system state gets updated. 
- Operators care about updates by implementing the update function. E.g. our ingress-operator. At the moment it is implemented in a weird way because back then we did not have the update support. What I think is more right in this case is this. The ingress-operator has a single resource configmap which gets updated over the course of its lifetime. Here the add function has to ensure this resource, the configmap exists. The update function has to ensure the configmap is in the right state. Here it makes sure the right ingress ports are always up to date.

**Comment by kopiczko** [comment]
My general thoughts about updating being unnecessary:

1. Because of ResyncPeriod and opeartor pod restarts the operator has to handle updates in `addFunc` anyway.
2. `onUpdate` function is just feature of the client library. It isn't a feature of  the K8s API. K8s API allows only watching and listing resources. The feature is also unreliable. Some updates may be skipped (see 3.).
3. When `onUpdate(oldObj, newObj)` is triggered, we cannot guarantee that `oldObj` is processed.

- - -

Also proven tools like Ansible or Chef follow that philosophy. They have have only create and delete state.

> There are indeed operators who do not care about updates

IMO in that case the operator should do nothing when subsequent `onAdd` is triggered and it detects managed resources are already created. This **must** be handled anyway because of the points above.

**Comment by xh3b4sd** [comment]
> Also proven tools like Ansible or Chef follow that philosophy.

This is a real argument for me. 

---

I would like to keep the separation of handling creation and updates of resources because of the example I gave. What about not using the k8s client's update func and calling the operatorkit interface update funcs after the creation parts within the add function?

**Comment by kopiczko** [comment]
> What about not using the k8s client's update func and calling the operatorkit interface update funcs after the creation parts within the add function?

1. How to implement `ResyncPeriod`?
2. How to implement operator startup with partially processed TPOs?
3. How to ensure that `oldObj` in `onUpdate(oldObj, newObj)` is already processed?

**Comment by xh3b4sd** [comment]
> How to implement ResyncPeriod?

I don't understand the question. What do we need to implement there? The resync period is configurable. We do not need to implement anything there. See https://github.com/giantswarm/operatorkit/blob/ebba313a2a521c384192e0011959d58682220cb3/tpr/tpr.go#L45. 

> How to implement operator startup with partially processed TPOs?

This is part of the "resource implementations have to be idempotent" thingy. IMO this has nothing to do with where to to execute the update logic. 

> How to ensure that oldObj in onUpdate(oldObj, newObj) is already processed?

What I suggested was to call [Framework. ProcessUpdate](https://github.com/giantswarm/operatorkit/blob/ebba313a2a521c384192e0011959d58682220cb3/framework/framework.go#L163) at the very end of [Framework. ProcessCreate](https://github.com/giantswarm/operatorkit/blob/ebba313a2a521c384192e0011959d58682220cb3/framework/framework.go#L96). There wouldn't be a difference then between old and new object. There is only the current object then.

**Comment by rossf7** [comment]
IMO calling ProcessUpdate at the end of ProcessCreate is a workable solution. It ensures that the resource is created before checking if the resources needs to be updated. I like that it means we don't need to implement the `updateFunc`.

> The update logic is more complex and needs to be implemented in the framework and the operators.

Yes this wasn't very clear and doesn't help the discussion. I agree only certain resources care about updates. For something like the Resource Group in Azure there is only create and delete. Implementing updates is optional so I think we're good here.

**Comment by kopiczko** [comment]
>> How to implement operator startup with partially processed TPOs?
>
> This is part of the "resource implementations have to be idempotent" thingy. IMO this has nothing to do with where to to execute the update logic.

For me updating a resource that doesn't exists is creating it. Updating partially created resource is finishing the creation (may also involve changing the resource). And finally updating the resource that is already created it's pure update. Since we have to deal with partially created resources it is all very mixed, and that's one of the reasons I don't want to distinguish between the two.

> What I suggested was to call Framework. ProcessUpdate at the very end of Framework. ProcessCreate. 

I missed that. As Ross said that would be a workable solution. There is one edge case I can think of though. Again, waht if `ProcessCreate` failed and subsequent `ProcessUpdate` is running on partially created resources? Can the framework user be sure that `ProcessUpdate` is called only if `ProcessCreate` was ran at least once?

**Comment by xh3b4sd** [comment]
> For me updating a resource that doesn't exists is creating it. Updating partially created resource is finishing the creation (may also involve changing the resource).

This is wrong. When nothing is there you have to create it. If nothing is there you cannot update something. When something is there you cannot create it. When something is there you have to update it. I want the creation and the update implementations separated because these are different things. I see the framework as something that guides you through the steps. When you want to implement everything in the same step, don't use the framework. 

> Again, waht if ProcessCreate failed and subsequent ProcessUpdate is running on partially created resources?

This cannot happen in what I suggest. The update functionality will only be called in case the creation did not fail. The creation either creates or does nothing. If it fails, the update will not be executed.

**Comment by kopiczko** [comment]
Let's say that operator reads the TPO and based on that it has to create two deployments: A and B.

## Scenario 1

1. The opeartor reads the TPO. Nothing is created.
2. The operator creates deployment A.
3. The operator fails on creating deployment B.
4. Either:
   a. After resync period the operator tires to create deployment B. (_This is "create"._)
   b. The TPO is updated. The operator tires to update deployment A and create deployment B. (_Is this "create" or "update"?_)

Point 4a. would trigger `onAdd` which in turn triggers `ProcessCreate` and it makes sense.
Point 4b. would trigger `onUpdate` which in turn triggers `ProcessUpdate` but it has to create the deployment which I guess could result in error.

## Scenario 2

1. The opeartor reads the TPO. Nothing is created.
2. The operator creates deployment A.
3. The operator creates deployment B.
4. The TPO is updated, forcing the operator to update deployment A.
5. The operator fails to process update event.
6. After resync  period the operator updates deployment A.

Point 6. would trigger `onAdd` which in turn triggers `ProcessCreate` even though it's a pure update.

- - -

Am I missing something there?

**Comment by xh3b4sd** [comment]
I think you miss that we are talking about not using `onUpdate` anymore at all. `ProcessCreate` and `ProcessUpdate` are called in this order in `onAdd` based on my initial suggestion here.

**Comment by kopiczko** [comment]
And `onUpdate(oldObj, newObj)` just calls `onAdd(newObj)` under the framework's hood?

**Comment by xh3b4sd** [comment]
Pawel, **we do not use `onUpdate`** in this proposal! See the very rough pseudo code below. 

We only register the add and delete functions. 
```
OnAdd: addFunc,
OnDelete: deleteFunc,
```

We call `ProcessCreate` and `ProcessUpdate` in the add function. 
```
func addFunc() {
  Framework.ProcessCreate()
  Framework.ProcessUpdate()
}
```

**Comment by kopiczko** [comment]
Ok, when we don't implement `OnUpdate` that makes sense 👍
I think we can delegate `OnUpdate` to the `OnAdd` inside the framework. But that's the detail.

**Comment by rossf7** [comment]
I've created giantswarm/giantswarm#1839 to implement this in our operators. Closing as I think we have consensus. Please reopen if we don't! :trollface:

**Comment by xh3b4sd** [comment]
I noticed a "weird" behaviour in the kvm-operator and investigated a bit. What I see in the logs of our operators is different to what people argued in the discussion we had here. 

- We presumed the order of the add function and the update function is not predictable. 
- We presumed the resync period applies to the add function as well so we have to update in there anyway. 

These assumptions are wrong based on my observations. The aws-operator, the kvm-operator and the draughtsman-operator I checked do the following. 

- On operator start the add function is executed. 
- On TPO creation the add function is executed.
- After the resync period the update function is executed. 
- The add function is not executed with the resync period. 
- The add function is always executed before the update function. 

I suggest to go back to my initial proposal. This means we do not need to do anything and https://github.com/giantswarm/giantswarm/issues/1839 can be closed because it is invalid.

**Comment by xh3b4sd** [comment]
I noticed this because the kvm-operator was not doing anything after the resync period. You can also check the logs of the aws-operator. There is only a massive amount of access logs of the health check route, but no logs from reconciliation based on the resync period. I also tested this by removing a service manifest from a KVM guest cluster. The kvm-operator does not implement yet the update functionality. Thus the service manifest was not recreated. Only after restarting the kvm-operator pod the missing service manifest was recreated because the add function was executed. On the other side the draughtsman-operator reconciles all the time, because it implements the update methods of the resource interface. When projects change they get corrected immediately by  the operator because the resync period triggers the updates.

**Comment by kopiczko** [comment]
Those 3 things seems to be in contradiction:

>* After the resync period the update function is executed.
>* The add function is not executed with the resync period.
>* The add function is always executed before the update function.

If `add` is always executed before `update`, we can forget about the `update` and proceed with https://github.com/giantswarm/giantswarm/issues/1839.

If only `update` is executed after resync period we should redirect `update` to `add` IMO.

I still think `ProcessCreate` should do updating, because of operator restarts and `add` is executed on updates. Therefore `ProcessUpdate` seems to be redundant.

**Comment by kopiczko** [comment]
Test case for that would be:
 
* stop the operator
* remove a service manifest
* start the operator and see if it recreated manifest before resync period

**Comment by xh3b4sd** [comment]
Cool, we had a quick brainstorming session where we agreed we are going with the following. 

```
{
  AddFunc: addFunc,
  DeleteFunc: deleteFunc,
  UpdateFunc: updateFunc,
}
```

```
func addFunc(obj) {
  ProcessCreate()
  ProcessUpdate()
}
```

```
func updateFunc(old, new) {
  addFunc(new)
}
```

The new issue is still valid. Closing this here. We know how to roll. Thanks gang.
